src/test/java/net/snowflake/client/jdbc/cloud/storage/EncryptionProviderTest.java

```diff
@@ -0,0 +1,101 @@
+package net.snowflake.client.jdbc.cloud.storage;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.security.SecureRandom;
+import java.util.Base64;
+import javax.crypto.CipherInputStream;
+import net.snowflake.client.jdbc.MatDesc;
+import net.snowflake.common.core.RemoteStoreFileEncryptionMaterial;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+public class EncryptionProviderTest {
+  private final SecureRandom random = new SecureRandom();
+
+  private final ArgumentCaptor<StorageObjectMetadata> storageObjectMetadataArgumentCaptor =
+      ArgumentCaptor.forClass(StorageObjectMetadata.class);
+  private final ArgumentCaptor<MatDesc> matDescArgumentCaptor =
+      ArgumentCaptor.forClass(MatDesc.class);
+  private final ArgumentCaptor<byte[]> ivDataArgumentCaptor = ArgumentCaptor.forClass(byte[].class);
+  private final ArgumentCaptor<byte[]> encKekArgumentCaptor = ArgumentCaptor.forClass(byte[].class);
+  private final ArgumentCaptor<Long> contentLengthArgumentCaptor =
+      ArgumentCaptor.forClass(Long.class);
+
+  private final StorageObjectMetadata meta = mock(StorageObjectMetadata.class);
+  private final SnowflakeStorageClient storageClient = mock(SnowflakeStorageClient.class);
+
+  private final String queryStageMasterKey =
+      Base64.getEncoder().encodeToString(random.generateSeed(16));
+  private final RemoteStoreFileEncryptionMaterial encMat = new RemoteStoreFileEncryptionMaterial();
+
+  byte[] plainText = "the quick brown fox jumps over the lazy dog".getBytes(StandardCharsets.UTF_8);
+
+  @Before
+  public void setUp() {
+    encMat.setQueryStageMasterKey(queryStageMasterKey);
+    encMat.setSmkId(123);
+    encMat.setQueryId("query-id");
+  }
+
```

Comment:
I'd like to also see some failure cases in these tests. eg., invalid ciphertext, valid ciphertext but invalid tag, etc

Reply:
Added some test cases with tampered data.